### PR TITLE
[io] Fix segfault in ReadObjectAny

### PR DESCRIPTION
**Pull request by mxxo**
Fixes a segfault that surfaced during #6008 (trying to merge `RNTuple` objects). 

`RNTuple` doesn't inherit from `TObject`, so `ReadObj` ends up passing a null pointer to `ReadObjectAny`. 
We add a null check to `ReadObjectAny` and return `0` if `ReadObj` is called on classes that don't inherit from `TObject`. 

Before: 
```shell
../bin/hadd -f merged.root ntuple1.root ntuple2.root
hadd Target file: merged.root
hadd compression setting for all output: 1
hadd Source file 1: ntuple1.root
hadd Source file 2: ntuple2.root
hadd Sources and Target have different compression levels
hadd merging will be slower
hadd Target path: merged.root:/

 *** Break *** segmentation violation
```

After:
```shell
 ../bin/hadd -f merged.root ntuple1.root ntuple2.root
hadd Target file: merged.root
hadd compression setting for all output: 1
hadd Source file 1: ntuple1.root
hadd Source file 2: ntuple2.root
hadd Sources and Target have different compression levels
hadd merging will be slower
hadd Target path: merged.root:/
Error in <TKey::ReadObj>: class ROOT::Experimental::RNTuple does not derive from TObject, call ReadObjectAny instead
Info in <TFileMerger::MergeRecursive>: could not read object for key {ntuple, }
```

### CODE REVIEW

**Comment by mxxo** [comment]
The test failures are related, I will look again at `ReadObjectAny` internals.
I think some of the failures are harmless (a new line added to the log, e.g.  `roottest-root-io-emulated-make` ), but some are perhaps real errors because I didn't call `ReadObjectAny` with `0` as an argument e.g. `roottest-root-io-cpp11Containers-unorderedSet`

File: io/io/src/TKey.cxx [hunk 1]
```diff
@@ -1014,9 +1015,13 @@ TObject *TKey::ReadObjWithBuffer(char *bufferRead)
 
 void *TKey::ReadObjectAny(const TClass* expectedClass)
 {
+   if (!expectedClass) {
```
Comment (pcanal) [inline io/io/src/TKey.cxx:1018]: This is a new limitation.  It is intended for this routine to have a 'no check mode'.

File: io/io/src/TKey.cxx [hunk 2]
```diff
@@ -756,7 +756,8 @@ TObject *TKey::ReadObj()
    }
    if (!cl->IsTObject()) {
       // in principle user should call TKey::ReadObjectAny!
-      return (TObject*)ReadObjectAny(0);
```
Comment (pcanal) [inline io/io/src/TKey.cxx:759]: This was necessary for backward compatibility with existing code with included:
```
MyType *obj = (MyType*)key->ReadObj();
```

**Comment by pcanal** [comment]
What is the stack trace of the hadd failure?

**Comment by mxxo** [comment]
No actual stack trace (will work on getting one), but traced crash to here: 
https://github.com/root-project/root/blob/ffc7c588ac91aca30e75d356ea971129ee6a836a/io/io/src/TFileMerger.cxx#L584-L586
where we dereference an invalid `hobj` pointer.

**Comment by pcanal** [comment]
Humm ... clearly that code need to be updated as it indeed "assume" that having a Merge function implies inheritance from TObject.

**Comment by pcanal** [comment]
It might be enough to use:
```
                        if (cl->IsTObject()) {
                           if (hobj->InheritsFrom(TCollection::Class())) {
                              ((TCollection*)hobj)->SetOwner();
                           }
                           hobj->ResetBit(kMustCleanup);
                       }
```

**Comment by mxxo** [comment]
Thank you for your help :). I just tried it, and it lets us call the merge function, but segfaults later on (perhaps during destruction when calling `inputs.Delete()`?). 
https://github.com/root-project/root/blob/ffc7c588ac91aca30e75d356ea971129ee6a836a/io/io/src/TFileMerger.cxx#L603-L609
`inputs` is `TList`, and `inputs.Add(obj)` takes a TObject pointer so I think that is the issue. 

`TCollection` is also expecting a `TObject`: 
https://github.com/root-project/root/blob/331efa4c00fefc38980eaaf7b41b8e95fcd1a23b/core/cont/inc/TCollection.h#L160

**Comment by pcanal** [comment]
> perhaps during destruction when calling inputs.Delete()

Yep, that wouldn't work.  Here we need to use instead of the Delete (for the non TObject case), an iteration over the content and calls to:
```
cl->Destructor( element_ptr );
```

**Comment by mxxo** [comment]
Closing because of the backwards compatibility break and I'll try again during the RNTuple merge implementation.

**Comment by mxxo** [comment]
@pcanal I think I resolved this particular segfault in 12fef499c109da4c2454d9cca2e60e88a883aa1c over at #6008.

**Comment by pcanal** [comment]
> @pcanal I think I resolved this particular segfault in 12fef49 over at #6008.

I am probably missing something but it looks like this commit would result in the RNTuple not actually being merged ... (because not add to the list and thus not passed to the merge function).

**Comment by mxxo** [comment]
> the RNTuple not actually being merged ... 

That is 100% correct, the actual merger is not implemented yet. I should have marked the PR `WIP` and made the RNTuple merger fail with an "unimplemented error message", but I wanted to highlight during execution that the attempted merger no longer segfaults.

**Pull request closed without merging**